deepchem/models/torch_models/cnn.py

```diff
@@ -0,0 +1,367 @@
+    self.residual = residual
+
+    n_layers = len(layer_filters)
+
+    # PyTorch layers require input and output channels as parameter
+    # if only one layer to make the model creating loop below work, multiply layer_filters wutg 2
+    if len(layer_filters) == 1:
+      layer_filters = layer_filters * 2
+
+    if not isinstance(kernel_size, list):
+      kernel_size = [kernel_size] * n_layers
+    if not isinstance(strides, SequenceCollection):
+      strides = [strides] * n_layers
+    if not isinstance(dropouts, SequenceCollection):
+      dropouts = [dropouts] * n_layers
+    if not isinstance(activation_fns, SequenceCollection):
+      activation_fns = [activation_fns] * n_layers
+
+    if uncertainty:
+
+      if mode != 'regression':
+        raise ValueError("Uncertainty is only supported in regression mode")
+
+      if any(d == 0.0 for d in dropouts):
+        raise ValueError(
+            'Dropout must be included in every layer to predict uncertainty')
+
+    # Python tuples use 0 based indexing, dims defines number of dimension for convolutional operation
+    ConvLayer = (nn.Conv1d, nn.Conv2d, nn.Conv3d)[self.dims - 1]
+
+    PoolLayer: Type
+    if pool_type == 'average':
+      PoolLayer = (nn.AvgPool1d, nn.AvgPool2d, nn.AvgPool3d)[self.dims - 1]
+    elif pool_type == 'max':
+      PoolLayer = (nn.MaxPool1d, nn.MaxPool2d, nn.MaxPool3d)[self.dims - 1]
+    else:
+      raise ValueError("pool_type must be either 'average' or 'max'")
+
+    self.layers = nn.ModuleList()
+
+    in_shape = n_features
+
+    for out_shape, size, stride, dropout, activation_fn in zip(
+        layer_filters, kernel_size, strides, dropouts, activation_fns):
+
+      self.layers.append(
+          ConvLayer(in_channels=in_shape,
+                    out_channels=out_shape,
+                    kernel_size=size,
+                    stride=stride,
+                    padding=padding,
+                    dilation=1,
+                    groups=1,
+                    bias=True))
+
+      if dropout > 0.0:
+        self.layers.append(nn.Dropout(dropout))
+
+      if activation_fn is not None:
+        self.layers.append(activation_fn())
+
+      self.layers.append(PoolLayer(size))
+      in_shape = out_shape
+
+  def forward(self, x: torch.Tensor) -> List[Any]:
+    """
+    Parameters
+    ----------
+    x: torch.Tensor
+      Input Tensor
+
+    Returns
+    -------
+    torch.Tensor
+      Output as per use case : regression/classification
+    """
+    prev_layer = x
+
+    for layer in self.layers:
+      x = layer(x)
+      # residual blocks can only be used when successive layers have the same output shape
+      if self.residual and layer.in_channels == layer.out_channels:
+        x = x + prev_layer
+      prev_layer = x
+
+    outputs = []
+    batch_size = x.shape[0]
+
+    x = torch.reshape(x, (batch_size, -1))
+
+    if self.mode == "classification":
+
+      logits = nn.Linear(x.shape[1], self.n_tasks * self.n_classes)(x)
```

Comment:
@peastman, Let's discuss about the solution to root issue, If we are to define this nn.Linear in constructor, how can we get the input shape which is x.shape[1] ?
pytorch defines graphs dynamically , which means we can get it only at runtime

I'm trying to replicate your [TF Implementation in torch](https://github.com/deepchem/deepchem/blob/66116f1b8edd8130eff0a435c8cb8d3deefaab75/deepchem/models/cnn.py#L205)